### ISSUE TITLE
Cannot use autotrack with AWS Kinesis

### ISSUE THREAD

**Issue by aldo-roman** [aws-amplify/amplify-js#4151]
**Describe the bug**
Cannot use autotrack with AWS Kinesis

**To Reproduce**

1. Setup Analytics with Kinesis
2. Setup session and/or pageview autotrack

**Expected behavior**
Sessions and/or Pageviews should appear in Kinesis dashboard

**Screenshots**

**Desktop (please complete the following information):**

- OS: All
- Browser: All
- Version: 1.2.25

**Additional context**
Works with AWS Pinpoint provider

```
[DEBUG] 55:07.394 AWSKineisProvider - configure Analytics {disabled: undefined}
ConsoleLogger.js:111 [DEBUG] 55:07.395 AnalyticsClass - configure Analytics {streamName: "test-stream", AWSKinesis: {…}}
ConsoleLogger.js:111 [DEBUG] 55:07.395 AWSKineisProvider - configure Analytics {disabled: undefined, streamName: "test-stream", region: "us-west-2", bufferSize: 1, flushSize: 1, …}
ConsoleLogger.js:111 [DEBUG] 55:07.396 Hub - Dispatching to analytics with  {event: "configured", data: null, message: "The Analytics category has been configured successfully"}
ConsoleLogger.js:111 [DEBUG] 55:07.396 Analytics - on hub capsule analytics {event: "configured", data: null, message: "The Analytics category has been configured successfully"}
ConsoleLogger.js:111 [DEBUG] 55:07.396 AnalyticsClass - current configuration {streamName: "test-stream", AWSKinesis: {…}, autoSessionRecord: true}
ConsoleLogger.js:111 [DEBUG] 55:10.316 AWSKineisProvider - set credentials for analytics undefined
ConsoleLogger.js:111 [DEBUG] 56:38.604 AWSKineisProvider - set credentials for analytics {accessKeyId: "...", sessionToken: "...", secretAccessKey: "...", identityId: "...", authenticated: true}
ConsoleLogger.js:111 [DEBUG] 56:38.604 AWSKineisProvider - set credentials for analytics {accessKeyId: "...", sessionToken: "...", secretAccessKey: "...", identityId: "...", authenticated: true}
ConsoleLogger.js:101 [DEBUG] 56:42.577 AWSKineisProvider - no change for analytics config, directly return from init
ConsoleLogger.js:101 [DEBUG] 59:00.122 AWSKineisProvider - no change for analytics config, directly return from init
```

**Comment by aldo-roman** [comment]
Kinesis provider expects an attribute `streamName` to be injected on *each* record.

https://github.com/aws-amplify/amplify-js/blob/9a1055a2eeaf15a1406032bf4c039137edd09fe1/packages/analytics/src/Providers/AWSKinesisProvider.ts#L175

A quick solution would be to specify a `streamName` on `Analytics.configure` and make it default for all records.

**Comment by elorzafe** [comment]
@aldo-roman can you try adding the pluggable doing configure? I think that can solve the problem meanwhile we find a fix for this.

**Comment by elorzafe** [comment]
> @aldo-roman can you try adding the plugin before doing `Amplify.configure`? I think that can solve the problem meanwhile we find a fix for this.

Dont consider that [comment](https://github.com/aws-amplify/amplify-js/issues/4151#issuecomment-542016806), this is what you should do
```
Analytics.autoTrack('event', {

    enable: true,
    // OPTIONAL, the attributes of the event, you can either pass an object or a function 
    // which allows you to define dynamic attributes
    attributes: {
        attr: 'attr'
    },
    // when using function
    // attributes: () => {
    //    const attr = somewhere();
    //    return {
    //        myAttr: attr
    //    }
    // },
    // OPTIONAL, the service provider, by default is the AWS Pinpoint
    provider: 'AWSKinesis'
});
```
[reference](https://aws-amplify.github.io/docs/js/analytics#session-tracking)

**Comment by aldo-roman** [comment]
@elorzafe yes, that is what I did. The autotrack events fire, but there is no way to indicate which AWS Kinesis stream it should report to. That's why I thought about using a default stream, but there can be other solutions.

**Comment by jordanranz** [comment]
It seems `Analytics.autoTrack` may need a `streamName` passed in to report to a specific stream. I'll mark this is a feature request.

**Comment by goatandsheep** [comment]
@jordanranz can you show an example that uses `streamName`?

**Comment by jordanranz** [comment]
@goatandsheep, unfortunately this functionality does not exist today.

**Comment by mauerbac** [comment]
Looks like this has become a Feature Request. Will close the issue out.

**Comment by aldo-roman** [comment]
@mauerbac Shouldn't we leave this open _precisely_ because it is a feature request?

**Comment by mauerbac** [comment]
Yes, sorry @aldo-roman. Will just remove the question label!

**Comment by goatandsheep** [comment]
@mauerbac are were closing out feature requests? Regardless, you should indicate on the docs that auto track doesn't work with kinesis for the time being

**Comment by njdullea** [comment]
Where does this feature request to fill in the development timeline? Do y'all expect it to be implemented in the near future?

**Comment by goatandsheep** [comment]
I think someone should make a pull request saying that `autotrack` does not work for Kinesis so people don't get confused as it seems like this very important feature is being overlooked.

**Comment by imaitland** [comment]
@mauerbac any action on this? Would like to be able to use autoTrack features with Kinesis.

**Comment by justro** [comment]
@mauerbac I've run into this too. Would you be willing to take a pull request for this?

What I was thinking is that I would add a field called 'eventParameters' to the page/session/event tracker options, e.g. https://github.com/aws-amplify/amplify-js/blob/9924a31397761fccd03f53336d754b98367199a8/packages/analytics/src/types/Analytics.ts#L34. This would be a `[key: string]: string;` collection similar to EventAttributes (https://github.com/aws-amplify/amplify-js/blob/9924a31397761fccd03f53336d754b98367199a8/packages/analytics/src/types/Analytics.ts#L26).

In the trackers, when `this._tracker` is called, each tracker would look for this collection in its config and take each param and inject it into the event, similar to the way eventName is handled.

This would allow users to configure a streamName and partitionKey that would be used by the trackers when recording events, e.g.

```
Analytics.autoTrack('pageView', {
  enable: true,
  type: 'SPA',
  provider: 'AWSKinesis',
  eventName: 'myKinesisPageViews',
  eventParameters: {
     streamName: 'my-kinesis-stream',
     partitionKey: 'my-partition-key'
  }
});
```

Looking at the existing code, it feels like putting this info on the tracker is the right thing to do, rather than for example making it like a default on the provider. This approach would allow users to send page/event/session autoTracking to different streams/partitions. Downside is it's a bit open-ended...IDK if it's valuable or troublesome to allow arbitrary fields on the events written by the tracker...I could see it being useful in other situations, but people might also do things like typo 'straemName' and waste time trying to figure out why their events aren't landing. On the other hand, it seems less right to encode a bunch of kinesis specific type info in the tracker options.

Anyway, willing to give this a shot if ya'll are interested.

**Comment by justro** [comment]
For anyone who's trying to make this work, after spending some time putting together a change to the amplify libraries, I realized all we need to do is create a simple wrapper around the AWSKinesisProivder, which we can use without modifying amplify-js because the library is designed to allow provider plugins. I was able to see events coming from my react-js app appearing live in a zeppelin notebook via the Kinesis Studio Notebooks using this approach.

My typescript implementation is here, tried to add lots of comments to explain usage: https://gist.github.com/justro/f72618744d96d9b26acedbb6a711f530
